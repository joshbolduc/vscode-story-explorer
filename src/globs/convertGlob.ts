import { relative } from 'path';
import type { Uri } from 'vscode';
import type { GlobSpecifier } from '../config/GlobSpecifier';
import { getGlobInfo } from './getGlobInfo';

interface ConvertedGlobWithoutPattern {
  globBase: Uri;
  globPattern?: never;
  filter?: never;
  regex?: never;
}

interface ConvertedGlobWithPattern {
  globBase: Uri;
  globPattern: string;
  filter: (uri: Uri) => boolean;
  regex: RegExp;
}

export type ConvertedGlob =
  | ConvertedGlobWithoutPattern
  | ConvertedGlobWithPattern;

// FUTURE: it may be possible to convert a subset of globs (where `isGlob` is
// true) to a VS Code-compatible form that doesn't require post-filtering
export const convertGlob = (globSpecifier: GlobSpecifier): ConvertedGlob => {
  const globInfo = getGlobInfo(globSpecifier);

  if (!globInfo.isGlob) {
    return {
      globBase: globInfo.base,
    };
  }

  const { base, regex } = globInfo;

  return {
    globBase: base,
    globPattern: '**/*',
    filter: (uri) => {
      if (base.scheme !== uri.scheme || base.authority !== uri.authority) {
        return false;
      }

      const relativePath = relative(base.path, uri.path);
      return regex.test(relativePath);
    },
    regex,
  };
};
