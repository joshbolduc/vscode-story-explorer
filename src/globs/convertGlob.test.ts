import { URI } from 'vscode-uri';
import { convertGlob } from './convertGlob';

describe('convertGlob', () => {
  const inputs = [
    ['file.ts'],
    ['src/file.ts'],
    [
      'src/*',
      ['/mock/baseDir/src/file.ts'],
      ['/mock/baseDir/file.ts', '/mock/baseDir/src/nested/file.ts'],
    ],
    [
      'src/**',
      [
        '/mock/baseDir/src/file.ts',
        '/mock/baseDir/src/nested/file.ts',
        '/mock/baseDir/src/file.js',
      ],
      ['/mock/baseDir/file.ts'],
    ],
    [
      'src/**/*.ts',
      ['/mock/baseDir/src/file.ts', '/mock/baseDir/src/nested/file.ts'],
      ['/mock/baseDir/src/file.js', '/mock/baseDir/file.ts'],
    ],
    [
      '../src/**/*.stories.mdx',
      ['/mock/src/test.stories.mdx', '/mock/src/nested/test.stories.mdx'],
      [
        '/mock/baseDir/src/test.stories.mdx',
        '/mock/src/stories.mdx',
        '/mock/src/nested/test.stories.mdx.txt',
      ],
    ],
    [
      '../src/**/*.stories.@(js|jsx|ts|tsx)',
      [
        '/mock/src/test.stories.js',
        '/mock/src/test.stories.jsx',
        '/mock/src/test.stories.ts',
        '/mock/src/test.stories.tsx',
        '/mock/src/nested/test.stories.ts',
      ],
      [
        '/mock/baseDir/src/test.stories.ts',
        '/mock/src/stories.ts',
        '/mock/src/nested/test.stories.ts.txt',
      ],
    ],
    [
      '**/*.stories.@(js|jsx|ts|tsx)',
      [
        '/mock/baseDir/src/test.stories.js',
        '/mock/baseDir/src/test.stories.jsx',
        '/mock/baseDir/src/test.stories.ts',
        '/mock/baseDir/src/test.stories.tsx',
        '/mock/baseDir/src/nested/test.stories.ts',
      ],
      [
        '/mock/src/test.stories.ts',
        '/mock/src/stories.ts',
        '/mock/src/nested/test.stories.ts.txt',
      ],
    ],
    [
      '**',
      [
        '/mock/baseDir',
        '/mock/baseDir/anything',
        '/mock/baseDir/nested/anything',
      ],
      ['/mock/src'],
    ],
    [
      '**/*',
      ['/mock/baseDir/anything', '/mock/baseDir/nested/anything'],
      ['/mock/src', '/mock/baseDir'],
    ],
    [
      '@(js|jsx|ts|tsx)',
      [
        '/mock/baseDir/js',
        '/mock/baseDir/jsx',
        '/mock/baseDir/ts',
        '/mock/baseDir/tsx',
      ],
      [
        '/mock/baseDir/file.js',
        '/mock/src/file.js',
        '/mock/baseDir/nested/file.js',
      ],
    ],
    [
      '**/../*.ts',
      ['/mock/test.ts'],
      [
        '/mock/baseDir/test.ts',
        '/mock/baseDir/nested/test.ts',
        '/mock/src/test.ts',
      ],
    ],
  ] as const;

  inputs.forEach(([input, match = undefined, noMatch = undefined]) =>
    it(`handles ${input}`, () => {
      const { filter, ...result } = convertGlob({
        directory: URI.file('/mock/baseDir'),
        files: input,
        titlePrefix: '',
      });

      if (!match) {
        expect(filter).toBeUndefined();
      } else {
        expect(typeof filter).toBe('function');
      }

      (match ?? []).forEach((path) => {
        expect(filter!(URI.file(path))).toBe(true);
      });
      (noMatch ?? []).forEach((path) => {
        expect(filter!(URI.file(path))).toBe(false);
      });

      const normalizedResult = {
        ...result,
        globBase: result.globBase.toString(),
      };

      expect(normalizedResult).toMatchSnapshot();
    }),
  );
});
