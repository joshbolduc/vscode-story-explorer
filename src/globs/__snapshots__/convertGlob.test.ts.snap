// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertGlob handles ** 1`] = `
{
  "globBase": "file:///mock/baseDir",
  "globPattern": "**/*",
  "regex": /\\^\\(\\?:\\(\\?:\\(\\?:\\(\\?!\\(\\?:\\^\\|\\\\/\\)\\\\\\.\\)\\.\\)\\*\\?\\)\\)\\$/,
}
`;

exports[`convertGlob handles **/* 1`] = `
{
  "globBase": "file:///mock/baseDir",
  "globPattern": "**/*",
  "regex": /\\^\\(\\?:\\(\\?:\\^\\|\\\\/\\|\\(\\?:\\(\\?:\\(\\?!\\(\\?:\\^\\|\\\\/\\)\\\\\\.\\)\\.\\)\\*\\?\\)\\\\/\\)\\(\\?!\\\\\\.\\)\\(\\?=\\.\\)\\[\\^/\\]\\*\\?\\\\/\\?\\)\\$/,
}
`;

exports[`convertGlob handles **/*.stories.@(js|jsx|ts|tsx) 1`] = `
{
  "globBase": "file:///mock/baseDir",
  "globPattern": "**/*",
  "regex": /\\^\\(\\?:\\(\\?:\\^\\|\\\\/\\|\\(\\?:\\(\\?:\\(\\?!\\(\\?:\\^\\|\\\\/\\)\\\\\\.\\)\\.\\)\\*\\?\\)\\\\/\\)\\(\\?!\\\\\\.\\)\\(\\?=\\.\\)\\[\\^/\\]\\*\\?\\\\\\.stories\\\\\\.\\(js\\|jsx\\|ts\\|tsx\\)\\)\\$/,
}
`;

exports[`convertGlob handles **/../*.ts 1`] = `
{
  "globBase": "file:///mock/baseDir",
  "globPattern": "**/*",
  "regex": /\\^\\(\\?:\\(\\?:\\^\\|\\\\/\\|\\(\\?:\\(\\?:\\(\\?!\\(\\?:\\^\\|\\\\/\\)\\\\\\.\\)\\.\\)\\*\\?\\)\\\\/\\)\\\\\\.\\\\\\.\\\\/\\(\\?!\\\\\\.\\)\\(\\?=\\.\\)\\[\\^/\\]\\*\\?\\\\\\.ts\\)\\$/,
}
`;

exports[`convertGlob handles ../src/**/*.stories.@(js|jsx|ts|tsx) 1`] = `
{
  "globBase": "file:///mock/src",
  "globPattern": "**/*",
  "regex": /\\^\\(\\?:\\(\\?:\\^\\|\\\\/\\|\\(\\?:\\(\\?:\\(\\?!\\(\\?:\\^\\|\\\\/\\)\\\\\\.\\)\\.\\)\\*\\?\\)\\\\/\\)\\(\\?!\\\\\\.\\)\\(\\?=\\.\\)\\[\\^/\\]\\*\\?\\\\\\.stories\\\\\\.\\(js\\|jsx\\|ts\\|tsx\\)\\)\\$/,
}
`;

exports[`convertGlob handles ../src/**/*.stories.mdx 1`] = `
{
  "globBase": "file:///mock/src",
  "globPattern": "**/*",
  "regex": /\\^\\(\\?:\\(\\?:\\^\\|\\\\/\\|\\(\\?:\\(\\?:\\(\\?!\\(\\?:\\^\\|\\\\/\\)\\\\\\.\\)\\.\\)\\*\\?\\)\\\\/\\)\\(\\?!\\\\\\.\\)\\(\\?=\\.\\)\\[\\^/\\]\\*\\?\\\\\\.stories\\\\\\.mdx\\)\\$/,
}
`;

exports[`convertGlob handles @(js|jsx|ts|tsx) 1`] = `
{
  "globBase": "file:///mock/baseDir",
  "globPattern": "**/*",
  "regex": /\\^\\(\\?:\\(js\\|jsx\\|ts\\|tsx\\)\\)\\$/,
}
`;

exports[`convertGlob handles file.ts 1`] = `
{
  "globBase": "file:///mock/baseDir/file.ts",
}
`;

exports[`convertGlob handles src/* 1`] = `
{
  "globBase": "file:///mock/baseDir/src",
  "globPattern": "**/*",
  "regex": /\\^\\(\\?:\\(\\?!\\\\\\.\\)\\(\\?=\\.\\)\\[\\^/\\]\\*\\?\\\\/\\?\\)\\$/,
}
`;

exports[`convertGlob handles src/** 1`] = `
{
  "globBase": "file:///mock/baseDir/src",
  "globPattern": "**/*",
  "regex": /\\^\\(\\?:\\(\\?:\\(\\?:\\(\\?!\\(\\?:\\^\\|\\\\/\\)\\\\\\.\\)\\.\\)\\*\\?\\)\\)\\$/,
}
`;

exports[`convertGlob handles src/**/*.ts 1`] = `
{
  "globBase": "file:///mock/baseDir/src",
  "globPattern": "**/*",
  "regex": /\\^\\(\\?:\\(\\?:\\^\\|\\\\/\\|\\(\\?:\\(\\?:\\(\\?!\\(\\?:\\^\\|\\\\/\\)\\\\\\.\\)\\.\\)\\*\\?\\)\\\\/\\)\\(\\?!\\\\\\.\\)\\(\\?=\\.\\)\\[\\^/\\]\\*\\?\\\\\\.ts\\)\\$/,
}
`;

exports[`convertGlob handles src/file.ts 1`] = `
{
  "globBase": "file:///mock/baseDir/src/file.ts",
}
`;
